cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,774 @@
+    staged_content.delete()
+    status.increment_completed_steps()
+
+    _create_migration_artifacts_incrementally(
+        root_migrated_node=root_migrated_node,
+        source=source,
+        migration=migration,
+        status=status,
+    )
+
+    block_migrations = ModulestoreBlockMigration.objects.filter(overall_migration=migration)
+    status.increment_completed_steps()
+
+    status.set_state(MigrationStep.FORWARDING.value)
+    if forward_source_to_target:
+        block_sources_to_block_migrations = {
+            block_migration.source: block_migration for block_migration in block_migrations
+        }
+        for block_source, block_migration in block_sources_to_block_migrations.items():
+            block_source.forwarded = block_migration
+            block_source.save()
+
+        source.forwarded = migration
+        source.save()
+    status.increment_completed_steps()
+
+    status.set_state(MigrationStep.POPULATING_COLLECTION.value)
+    if target_collection:
+        block_target_pks: list[int] = list(
+            ModulestoreBlockMigration.objects.filter(
+                overall_migration=migration
+            ).values_list('target_id', flat=True)
+        )
+        if block_target_pks:
+            authoring_api.add_to_collection(
+                learning_package_id=target_package_pk,
+                key=target_collection.key,
+                entities_qset=PublishableEntity.objects.filter(id__in=block_target_pks),
+                created_by=user_id,
+            )
+        else:
+            log.warning("No target entities found to add to collection")
+    status.increment_completed_steps()
+
+
+@dataclass(frozen=True)
+class _MigratedNode:
+    """
+    A node in the source tree, its target (if migrated), and any migrated children.
+
+    Note that target_version can equal None even when there migrated children.
+    This happens, particularly, if the node is above the requested composition level
+    but has descendents which are at or below that level.
+    """
+    source_to_target: tuple[UsageKey, PublishableEntityVersion] | None
+    children: list[_MigratedNode]
+
+    def all_source_to_target_pairs(self) -> t.Iterable[tuple[UsageKey, PublishableEntityVersion]]:
+        """
+        Get all source_key->target_ver pairs via a pre-order traversal.
+        """
+        if self.source_to_target:
+            yield self.source_to_target
+        for child in self.children:
+            yield from child.all_source_to_target_pairs()
+
+
+def _migrate_node(
+    *,
+    context: _MigrationContext,
+    source_node: XmlTree,
+) -> _MigratedNode:
+    """
+    Migrate an OLX node (source_node) from a legacy course or library (source_context_key) to a
+    learning package (context.target_library). If the node is a container, create it in the target if
+    it is at or above the requested composition_level; otherwise, just import its contents.
+    Recursively apply the same logic to all children.
```

Comment:
source_context_key -> context.source_context_key
```suggestion
    Migrate an OLX node (source_node) from a legacy course or library (context.source_context_key)
    to a learning package (context.target_library). If the node is a container, create it in the
    target if it is at or above the requested composition_level; otherwise, just import its contents.
    Recursively apply the same logic to all children.
```